editor/src/components/navigator/navigator-item/component-picker.tsx

```diff
@@ -0,0 +1,295 @@
+/** @jsxRuntime classic */
+/** @jsx jsx */
+import { jsx } from '@emotion/react'
+import React from 'react'
+import { useColorTheme } from '../../../uuiui'
+import { capitalize } from '../../../core/shared/string-utils'
+import { type PreferredChildComponentDescriptor } from '../../custom-code/internal-property-controls'
+import { type JSXElementChild } from '../../../core/shared/element-template'
+import { type Imports } from '../../../core/shared/project-file-types'
+import { elementFromInsertMenuItem } from '../../editor/insert-callbacks'
+
+export interface ComponentPickerProps {
+  insertionTargetName: string
+  preferredComponents: PreferredChildComponentDescriptor[]
+  allComponents: PreferredChildComponentDescriptor[]
+  onItemClick: (elementToInsert: ElementToInsert) => React.MouseEventHandler
+  onClickCloseButton?: React.MouseEventHandler
+}
+
+export interface ElementToInsert {
+  elementToInsert: (uid: string) => JSXElementChild
+  additionalImports: Imports
+}
+
+export function componentPickerTestIdForProp(prop: string): string {
+  return `component-picker-${prop}`
+}
+
+export const componentPickerCloseButtonTestId = `component-picker-close-button`
+export const componentPickerFilterInputTestId = `component-picker-filter-input`
+
+export function componentPickerOptionTestId(componentName: string, variant?: string): string {
+  const variantSuffix = variant == null ? '' : `-${variant}`
+  return `component-picker-option-${componentName}${variantSuffix}`
+}
+
+export const ComponentPicker = React.memo((props: ComponentPickerProps) => {
+  const colorTheme = useColorTheme()
+  const [selectedTab, setSelectedTab] = React.useState<'preferred' | 'all'>('preferred')
+  const switchToPreferredTab = React.useCallback(() => setSelectedTab('preferred'), [])
+  const switchToAllTab = React.useCallback(() => setSelectedTab('all'), [])
+  const [filter, setFilter] = React.useState<string>('')
+  const handleFilterKeydown = React.useCallback((e: React.KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'Enter') {
+      ;(e.target as HTMLInputElement).blur()
+    } else if (e.key === 'Escape') {
+      setFilter('')
+      ;(e.target as HTMLInputElement).blur()
+    }
+    e.stopPropagation()
+  }, [])
+  const handleFilterChange = React.useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
+    setFilter(e.target.value)
+  }, [])
+
+  const unfilteredComponentsToShow =
+    selectedTab === 'preferred' ? props.preferredComponents : props.allComponents
+  const componentsToShow =
+    filter.trim() === ''
+      ? unfilteredComponentsToShow
+      : unfilteredComponentsToShow.filter((v) =>
+          v.name.toLocaleLowerCase().includes(filter.toLocaleLowerCase().trim()),
+        )
+
+  return (
```

Comment:
In my opinion this returned jsx is way too long for a single component. Especially because there are good candidates in it which could be extracted into components, e.g. the view of the components themselves, maybe even the variants, or all the components under the search field together.

Reply:
Yeah, that's fair. I have a couple of follow up tasks for another PR I'll be submitting later today, so I'll do a bit of refactoring here as part of that too